pkg/distributor/distributor.go

```diff
@@ -688,7 +685,57 @@ func (d *Distributor) prePushHaDedupeMiddleware(next push.Func) push.Func {
 			d.nonHASamples.WithLabelValues(userID).Add(float64(numSamples))
 		}
 
-		cleanupInDefer = false
+		dontCleanup()
+		return next(ctx, req, cleanup)
+	}
+}
+
+func (d *Distributor) prePushRelabelMiddleware(next push.Func) push.Func {
+	return func(ctx context.Context, req *mimirpb.WriteRequest, cleanup func()) (*mimirpb.WriteResponse, error) {
+		conditionalCleanup, dontCleanup := getConditionalCleanup(cleanup)
+		defer conditionalCleanup()
+
+		userID, err := tenant.TenantID(ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		var removeTsIndexes []int
+		for tsIdx := 0; tsIdx < len(req.Timeseries); tsIdx++ {
+			ts := req.Timeseries[tsIdx]
+
+			if mrc := d.limits.MetricRelabelConfigs(userID); len(mrc) > 0 {
+				l := relabel.Process(mimirpb.FromLabelAdaptersToLabels(ts.Labels), mrc...)
+				ts.Labels = mimirpb.FromLabelsToLabelAdapters(l)
+			}
+
+			for _, labelName := range d.limits.DropLabels(userID) {
+				removeLabel(labelName, &ts.Labels)
+			}
+
+			if len(ts.Labels) == 0 {
+				if len(removeTsIndexes) == 0 {
+					// If we have to add one index into this slice then there is a good chance that we'll have to add more.
+					removeTsIndexes = make([]int, 0, len(req.Timeseries))
+				}
+				removeTsIndexes = append(removeTsIndexes, tsIdx)
+				continue
+			}
+
+			// We rely on sorted labels in different places:
+			// 1) When computing token for labels, and sharding by all labels. Here different order of labels returns
+			// different tokens, which is bad.
+			// 2) In validation code, when checking for duplicate label names. As duplicate label names are rejected
+			// later in the validation phase, we ignore them here.
+			// 3) Ingesters expect labels to be sorted in the Push request.
+			sortLabelsIfNeeded(ts.Labels)
+		}
+
+		if len(removeTsIndexes) > 0 {
+			req.Timeseries, _, _ = util.RemoveSliceIndexes(req.Timeseries, removeTsIndexes)
+		}
+
```

Comment:
This is an optimization, while this PR should focus on refactor in the first place. I'd recommend to move optimizations to different PR, with a benchmark.

Reply:
I guess it could be called an optimization, I figured it's better if each middleware guarantees that its output is "clean" to prevent unexpected side effects if middlewares get combined differently in the future. Passing on time series without any labels seemed not clean.

Reply:
By "optimization" I meant using batch-removal, not original removal that we had before.

Reply:
ah, ok. that's a fair point.